File: src/main/java/com/market/ddanggyeo/jwt/JwtFilter.java
package com.market.ddanggyeo.jwt;

import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@Component
public class JwtFilter extends OncePerRequestFilter {
    private final JwtUtils jwtUtils;

    JwtFilter(JwtUtils jwtUtils) {
        this.jwtUtils = jwtUtils;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request,
                                    HttpServletResponse response,
                                    FilterChain filterChain) throws ServletException, IOException {
        String authorizationHeader = request.getHeader("Authorization");
        if (authorizationHeader == null) {
            filterChain.doFilter(request, response);
        }

        String token = authorizationHeader.substring("Bearer ".length());
        if (token == null) {
            filterChain.doFilter(request, response);
        }

        if (jwtUtils.validation(token)) {
            String email = jwtUtils.parseEmail(token);
            Authentication authentication = jwtUtils.getAuthentication(email);
            SecurityContextHolder.getContext().setAuthentication(authentication);
        }

        filterChain.doFilter(request, response);
    }
}

File: src/main/kotlin/com/market/ddanggyeo/jwt/JwtFilter.kt
package com.market.ddanggyeo.jwt

import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class JwtFilter(private val jwtUtils: JwtUtils): OncePerRequestFilter() {

    //token 으로 권한을 체크하고 authentication 을 생성해서 security 인증 필터를 통과하도록 조치
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val authorizationHeader: String? = request.getHeader("Authorization")?:
            return filterChain.doFilter(request, response)
        val token: String = authorizationHeader?.substring("Bearer ".length)?:
            return filterChain.doFilter(request, response)

        if (jwtUtils.validation(token)) {
            val email = jwtUtils.parseEmail(token)
            val authentication: Authentication = jwtUtils.getAuthentication(email)
            SecurityContextHolder.getContext().authentication = authentication
        }

        filterChain.doFilter(request, response)
    }

}